### ISSUE TITLE
ENH: Add `fail_if` - `backend_has_changed`

### ISSUE THREAD

**Issue by hmgaudecker** [iza-institute-of-labor-economics/gettsim#966]
In #960, we had a [subtle bug](https://github.com/iza-institute-of-labor-economics/gettsim/pull/960#issuecomment-2967777523) that used numpy instead of jax because in setting up the policy environment, we had set `backend="numpy"` so that `policy_environment["xnp"]` was numpy, which is picked up by some things (like `reorder_ids`), whereas we set  `interface_dag["backend"]="jax"` when running the tests.

This may happen in general and we should have a check for discrepancies between the same elements in both collections / dags. AFAICS, these are the following right now:

- `backend`
- `xnp`
- `dnp`
- `evaluation_date`

**Comment by hmgaudecker** [comment]
@MImmesberger, I added tests on the `fail-if-backend-has-changed` branch. However, I am a little bit of a loss regarding the strategy since
- we don't store the backend with the objects (and if we did, users could still put a numpy array into some env that should only have jax arrays)
- the environments hold very heterogeneous objects, so we can't do a check whitelisting allowed types.

Maybe the best thing would be to 
- catch the tracer error we are getting in `test_backend_has_changed_from_numpy_to_jax` and give a meaningful message? *(likely happens in `raw_results.columns`)*
- work around the type failures by allowing a union of `xnp`, `numpy`, and `jax.numpy` (if installed) in our tests?

**Comment by hmgaudecker** [comment]
> > the environments hold very heterogeneous objects, so we can't do a check whitelisting allowed types.
> 
> But we could do the reverse, right? Given a backend, there are some types we are sure we don't want in there. Doesn't need to be complete I think

See, that is why I am asking clever people! If you find time later to give it a shot, would be great!

> > work around the type failures by allowing a union of xnp, numpy, and jax.numpy (if installed) in our tests?
> 
> I don't get this. Could you elaborate?

Did you run the tests? Looks like `numpy` can handle `jax.numpy.array`s just fine, but we fail because the pandas converter does not receive a `numpy.array`.
